### PR TITLE
[Torch Dialect] Add support for AtenScalarTensorOp

### PR DESCRIPTION
**Pull request by zhekunz2**
as title: this PR adds support and decomposition for `aten.scalar_tensor`.

### CODE REVIEW

**Comment by zhekunz2** [comment]
Depending on https://github.com/llvm/torch-mlir/pull/2062

File: lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp [hunk 1]
```diff
@@ -121,6 +121,27 @@ class RecomposeSelectFill_ : public OpRewritePattern<AtenFill_TensorOp> {
...
+    // prim.NumToTensor.Scalar + aten.to.dtype_layout
+    Type resultTy = op.getResult().getType();
+    Value numToTensor = rewriter.create<PrimNumToTensorScalarOp>(
+        op.getLoc(), resultTy, op.getS());
+
+    Value cstNone = rewriter.create<ConstantNoneOp>(op.getLoc());
+    Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(op.getLoc(), false);
+    Value toDTypeLayout = rewriter.create<AtenToDtypeLayoutOp>(
+        op.getLoc(), resultTy, numToTensor, op.getDtype(), op.getLayout(),
+        op.getDevice(), op.getPinMemory(), cstFalse, cstFalse, cstNone);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp:140]: nit: add `/*varname=*/var` comments for `cstFalse` and `cstNone`

File: lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp [hunk 2]
```diff
@@ -121,6 +121,27 @@ class RecomposeSelectFill_ : public OpRewritePattern<AtenFill_TensorOp> {
...
   }
 };
+
+class RecomposeScalarTensor : public OpRewritePattern<AtenScalarTensorOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenScalarTensorOp op,
+                                PatternRewriter &rewriter) const override {
+    // rewrite aten.scalar_tensor to
+    // prim.NumToTensor.Scalar + aten.to.dtype_layout
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/RecomposeComplexOps.cpp:131]: This is a decomposition, so it should be in `DecomposeComplexOps`. The pattern in this file is for matching on a *series* of ops and (usually) collapsing it to a smaller series of ops that allow further optimizations to take place in torch-mlir.

**Review by ramiro050 — CHANGES_REQUESTED** [review]
Can you also add the corresponding shape and dtype functions?

File: python/torch_mlir_e2e_test/test_suite/basic.py [hunk 3]
```diff
@@ -3821,6 +3821,49 @@ def ConstantBoolParameterModule_basic(module, tu: TestUtils):
 # ==============================================================================
 
 
+class ScalarTensorFloat32Module(torch.nn.Module):
```
Comment (ramiro050) [inline python/torch_mlir_e2e_test/test_suite/basic.py:3824]: Can you add a float test without the `dtype` argument to test the default dtype path? PyTorch is a bit weird with how it turns floats to tensors, sometimes they turn into f32, and sometimes into f64

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py [hunk 4]
```diff
@@ -752,6 +752,15 @@ def aten〇tensor〇int〡shape(t: int, dtype: Optional[int] = None, device: Opt
 def aten〇tensor〇bool〡shape(t: bool, dtype: Optional[int] = None, device: Optional[device] = None, requires_grad: bool = False) -> List[int]:
     return []
 
+def aten〇scalar_tensor〡shape(s: float, dtype: Optional[int] = None, layout: Optional[int] = None, device: Optional[device] = None, pin_memory: Optional[bool] = None) -> List[int]:
+    return []
+
+def aten〇scalar_tensor〡dtype(s: Union[int, float], dtype: Optional[int] = None, layout: Optional[int] = None, device: Optional[device] = None, pin_memory: Optional[bool] = None) -> int:
```
Comment (ramiro050) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py:758]: Can you add tests to this function? Something like:

https://github.com/llvm/torch-mlir/blob/9c6a58b784df222970b176b28dd27938de7d2859/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/abstract_interp_lib_gen.py#L1842

to test both possible dtype values for `s`.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 5]
```diff
@@ -4356,6 +4356,34 @@ class DecomposeAtenVarMeanDimOp : public OpRewritePattern<AtenVarMeanDimOp> {
...
+        op.getLoc(),
+        resultTy.getWithSizesAndDtype(resultTy.getOptionalSizes(), scalarTy),
+        op.getS());
+
+    Value cstNone = rewriter.create<ConstantNoneOp>(op.getLoc());
+    Value cstFalse = rewriter.create<Torch::ConstantBoolOp>(op.getLoc(), false);
+    Value toDTypeLayout = rewriter.create<AtenToDtypeLayoutOp>(
+        op.getLoc(), resultTy, numToTensor, op.getDtype(), op.getLayout(),
+        op.getDevice(), op.getPinMemory(), /*non_blocking*/ cstFalse,
+        /*copy*/ cstFalse, /*memory_format*/ cstNone);
```
Comment (ramiro050) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:4380]: nit: format is `/*var_name=*/var` (missing `=` and remove space)

**Review by ramiro050 — CHANGES_REQUESTED** [review]

**Review by ramiro050 — APPROVED** [review]

**Review by qingyunqu — APPROVED** [review]
LGTM

**Pull request merged**